fair/forward.py

```diff
@@ -1,326 +1,309 @@
 
-  # Update the thermal response boxes
-  T_j[0] = RF[0,np.newaxis]*q*(1-np.exp((-tstep)/d)) + T_j_pre*np.exp(-tstep/d)
+    # Stratospheric water vapour is a function of the methane radiative forcing
+    F[0,6] = h2o_st.linear(F[0,1])
 
-  # Sum the thermal response boxes to get the total temperature anomlay
-  T[0] = np.sum(T_j[0])
+    # Forcing from contrails. As with tr O3, no feedback dependence
+    F[:,7] = contrails.from_aviNOx(emissions, aviNOx_frac)
 
-  # # # ------------ REST OF RUN ------------ # # #
-  for x in range(1,integ_len):
-    if conc_driven:
-      C[x] = co2_concs[x]
-    
+    # Forcing from aerosols - again no feedback dependence
+    if useStevens:
+      F[:,8] = aerosols.Stevens(emissions)
     else:
-      # Calculate the parametrised iIRF and check if it is over the maximum 
-      # allowed value
-      iirf100[x] = r0 + rC*C_acc[x-1] + rT*T[x-1]
-      if iirf100[x] >= iirf100_max:
-        iirf100[x] = iirf100_max
-        
-      # Determine a solution for alpha using scipy's root finder
-      time_scale_sf = (root(iirf100_interp_funct,time_scale_sf,args=(a,tau,iirf100[x])))['x']
-
-      # Multiply default timescales by scale factor
-      tau_new = time_scale_sf * tau
-
-      # Compute the updated concentrations box anomalies from the decay of the previous year and the emisisons
-      R_i[x] = R_i[x-1]*np.exp(-tstep/tau_new) \
-              + (emissions[x,np.newaxis])*a*tau_new*(1-np.exp(-tstep/tau_new)) / ppm_gtc
+      F[:,8] = aerosols.regress(emissions)
 
-      # Sum the boxes to get the total concentration anomaly
-      C[x] = np.sum(R_i[x]) + C_0
+    # Black carbon on snow - no feedback dependence
+    F[:,9] = bc_snow.linear(emissions)
 
-      # Calculate the additional carbon uptake
-      C_acc[x] =  C_acc[x-1] + emissions[x] * tstep - (C[x]-C[x-1]) * ppm_gtc
+    # Land use change - scales fairly well with cumulative land use C emissions.
+    # We assume no feedbacks from the carbon cycle. Perhaps future improvement.
+    F[:,10] = landuse.cumulative(emissions)
 
-    # Calculate the radiative forcing using the previous timestep's CO2 concentration
-    RF[x] = (F_2x/np.log(2.)) * np.log((C[x-1]) /C_0) + other_rf[x]
+    # Volcanic and solar copied straight to the output arrays
+    F[:,11] = F_volcanic
+    F[:,12] = F_solar
 
-    # Update the thermal response boxes
-    T_j[x] = T_j[x-1]*np.exp(-tstep/d) + RF[x,np.newaxis]*q*(1-np.exp(-tstep/d))
-    
-    # Sum the thermal response boxes to get the total temperature anomaly
-    T[x] = np.sum(T_j[x])
+    # multiply by scale factors: note does not apply for CO2, scale CO with F2x
+    F[0,1:] = F[0,1:] * scale[1:]
 
-  if restart_out:
-    return C, T, (R_i[-1],T_j[-1],C_acc[-1])
   else:
-    return C, T
-
-# Define a function to plot FAIR's inputs and outputs
-# Shifts variables to appropriately represent their definitions (continuous 
-# throughout timestep or end of timestep)
-def plot_fair(emms,
```

Comment:
I'm guessing none of this works with your set up?

Reply:
I could make it work, but I reckon it's better to just provide the user with the raw data and let them handle the plotting themselves. One option would be to make a suite of ancillary tools which would include a plotting function. The ensemble generator (I used R, but can translate using Rpy) would be another.

Reply:
Makes sense.

I really like the idea of ancillary tools. I found getting plotting going such a faff the first time (fiddling axis labels, getting units, making sensible titles, having legends which include useful information) that having an example of how to do it in the ancillary code file and a way to quickly see what the model is actually doing, before you start grinding through the code, would be super useful I think.